File: src/CannabisMath.Core/Weights/WeightConverter.cs
namespace CannabisMath.Core.Weights;

public static class WeightConverter
{
    public const decimal GramsPerRetailEighth = 3.5m;
    public const decimal GramsPerRetailQuarter = 7m;
    public const decimal GramsPerRetailHalfOunce = 14m;
    public const decimal GramsPerRetailOunce = 28m;

    public const decimal GramsPerExactOunce = 28.349523125m;
    public const decimal GramsPerExactPound = 453.59237m;

    public static decimal GramsToRetailEighths(decimal grams)
    {
        ValidateNonNegative(grams, nameof(grams));
        return grams / GramsPerRetailEighth;
    }

    public static decimal RetailEighthsToGrams(decimal eighths)
    {
        ValidateNonNegative(eighths, nameof(eighths));
        return eighths * GramsPerRetailEighth;
    }

    public static decimal GramsToRetailOunces(decimal grams)
    {
        ValidateNonNegative(grams, nameof(grams));
        return grams / GramsPerRetailOunce;
    }

    public static decimal RetailOuncesToGrams(decimal ounces)
    {
        ValidateNonNegative(ounces, nameof(ounces));
        return ounces * GramsPerRetailOunce;
    }

    public static decimal GramsToExactOunces(decimal grams)
    {
        ValidateNonNegative(grams, nameof(grams));
        return grams / GramsPerExactOunce;
    }

    public static decimal ExactOuncesToGrams(decimal ounces)
    {
        ValidateNonNegative(ounces, nameof(ounces));
        return ounces * GramsPerExactOunce;
    }

    public static decimal GramsToExactPounds(decimal grams)
    {
        ValidateNonNegative(grams, nameof(grams));
        return grams / GramsPerExactPound;
    }

    public static decimal ExactPoundsToGrams(decimal pounds)
    {
        ValidateNonNegative(pounds, nameof(pounds));
        return pounds * GramsPerExactPound;
    }

    private static void ValidateNonNegative(decimal value, string parameterName)
    {
        if (value < 0)
        {
            throw new ArgumentOutOfRangeException(parameterName, "Value cannot be negative.");
        }
    }
}

File: js/src/weights/weightConverter.ts
export class WeightConverter {
  static readonly GRAMS_PER_RETAIL_EIGHTH = 3.5;
  static readonly GRAMS_PER_RETAIL_QUARTER = 7;
  static readonly GRAMS_PER_RETAIL_HALF_OUNCE = 14;
  static readonly GRAMS_PER_RETAIL_OUNCE = 28;

  static readonly GRAMS_PER_EXACT_OUNCE = 28.349523125;
  static readonly GRAMS_PER_EXACT_POUND = 453.59237;

  static gramsToRetailEighths(grams: number): number {
    this.validateNonNegative(grams, "grams");
    return grams / this.GRAMS_PER_RETAIL_EIGHTH;
  }

  static retailEighthsToGrams(eighths: number): number {
    this.validateNonNegative(eighths, "eighths");
    return eighths * this.GRAMS_PER_RETAIL_EIGHTH;
  }

  static gramsToRetailOunces(grams: number): number {
    this.validateNonNegative(grams, "grams");
    return grams / this.GRAMS_PER_RETAIL_OUNCE;
  }

  static retailOuncesToGrams(ounces: number): number {
    this.validateNonNegative(ounces, "ounces");
    return ounces * this.GRAMS_PER_RETAIL_OUNCE;
  }

  static gramsToExactOunces(grams: number): number {
    this.validateNonNegative(grams, "grams");
    return grams / this.GRAMS_PER_EXACT_OUNCE;
  }

  static exactOuncesToGrams(ounces: number): number {
    this.validateNonNegative(ounces, "ounces");
    return ounces * this.GRAMS_PER_EXACT_OUNCE;
  }

  static gramsToExactPounds(grams: number): number {
    this.validateNonNegative(grams, "grams");
    return grams / this.GRAMS_PER_EXACT_POUND;
  }

  static exactPoundsToGrams(pounds: number): number {
    this.validateNonNegative(pounds, "pounds");
    return pounds * this.GRAMS_PER_EXACT_POUND;
  }

  private static validateNonNegative(value: number, name: string): void {
    if (value < 0) {
      throw new Error(`${name} cannot be negative`);
    }
  }
}